### PR TITLE
chore: prepare LS as CLI Extension

### PR DESCRIPTION
**Pull request by PeterSchafer**
### Description

* Introduce CLI Extension entry point in parallel to the existing entry point in main to remain existing behaviour.
* Introduce execution mode (standalone/extension) to differentiate
  * automatic binary management 
  * CLI Executor

### Checklist

- [x] Tests added and all succeed
- [x] Linted
- [x] README.md updated, if user-facing
- [ ] License file updated, if new 3rd-party dependency is introduced

### CODE REVIEW

File: application/config/config.go [hunk 1]
```diff
@@ -626,6 +629,13 @@ func (c *Config) SetManageBinariesAutomatically(enabled bool) {
 	c.manageBinariesAutomatically.Set(enabled)
 }
 
+func (c *Config) ManageCliBinariesAutomatically() bool {
+	if c.engine.GetConfiguration().GetString(cli_constants.EXECUTION_MODE_KEY) == cli_constants.EXECUTION_MODE_VALUE_EXTENSION {
```
Comment (bastiandoetsch) [inline application/config/config.go:633]: Why do we need `EXECUTION_MODE_VALUE_EXTENSION? Can't we use `EXECUTION_MODE_VALUE_STANDALONE`? The LS doesn't need to know if there is an extension mode.

File: infrastructure/cli/cli_constants/constants.go [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cli_constants
+
+const (
+	EXECUTION_MODE_KEY              string = "execution-mode"
+	EXECUTION_MODE_VALUE_EXTENSION  string = "extension"
```
Comment (bastiandoetsch) [inline infrastructure/cli/cli_constants/constants.go:21]: Could this go away?

File: infrastructure/cli/cli_extension_executor_test.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+	"github.com/snyk/go-application-framework/pkg/configuration"
+	"github.com/snyk/go-application-framework/pkg/workflow"
+	"github.com/spf13/pflag"
+	"github.com/stretchr/testify/assert"
+
+	"github.com/snyk/snyk-ls/application/config"
+)
+
+func Test_ExecuteLegacyCLI_SUCCESS(t *testing.T) {
+
```
Comment (bastiandoetsch) [inline infrastructure/cli/cli_extension_executor_test.go:33]: Is this a unit, integration or smoke test? Could you maybe add the appropriate testutil.IntegTest or such marker?

File: ls_extension/main.go [hunk 4]
```diff
@@ -0,0 +1,117 @@
+/*
```
Comment (bastiandoetsch) [inline ls_extension/main.go:1]: could we share code between main.go and this file?

File: ls_extension/main.go [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+) (output []workflow.Data, err error) {
+	output = []workflow.Data{}
+
+	c := config.CurrentConfig()
+	zerolog.SetGlobalLevel(zerolog.InfoLevel)
+
+	logger := invocation.GetEnhancedLogger()
+	extensionConfig := invocation.GetConfiguration()
+
+	logger.Log().Msgf("Language Server here!")
```
Comment (bastiandoetsch) [inline ls_extension/main.go:79]: Maybe just output "Language Server version: %s" or such?

Comment (bastiandoetsch) [reply ls_extension/main.go:79]: Also, most likely this should go to stderr - is that ensured?

**Review by bastiandoetsch — APPROVED** [review]

File: ls_extension/main.go (hunk 5 above)
Comment (PeterSchafer) [reply ls_extension/main.go:79]: Makes sense I'll adapt the text. This logger is safe to use though as it is already initialized by the engine before invoking the extension.

File: ls_extension/main.go (hunk 4 above)
Comment (PeterSchafer) [reply ls_extension/main.go:1]: add entrypoint helpers, to at least share some snippets.

File: infrastructure/cli/cli_constants/constants.go [hunk 6]
```diff
@@ -0,0 +1,23 @@
+/*
+ * © 2022 Snyk Limited All rights reserved.
```
Comment (bastiandoetsch) [inline infrastructure/cli/cli_constants/constants.go:2]: This should be 2023 ;)

File: infrastructure/cli/cli_extension_executor_test.go [hunk 7]
```diff
@@ -0,0 +1,92 @@
+/*
+ * © 2022 Snyk Limited All rights reserved.
```
Comment (bastiandoetsch) [inline infrastructure/cli/cli_extension_executor_test.go:2]: This, too *duckandrun*

File: infrastructure/cli/initializer.go [hunk 8]
```diff
@@ -66,7 +66,7 @@ func (i *Initializer) Init() error {
 	logger := log.With().Str("method", "cli.Init").Logger()
 	cliInstalled := config.CurrentConfig().CliSettings().Installed()
 	logger.Debug().Str("cliPath", cliPathInConfig()).Msgf("CLI installed: %v", cliInstalled)
-	if !config.CurrentConfig().ManageBinariesAutomatically() {
+	if !config.CurrentConfig().ManageCliBinariesAutomatically() {
```
Comment (bastiandoetsch) [inline infrastructure/cli/initializer.go:69]: This is very nitpicky of me. But it is only one binary. So it should most likely be `ManageCliBinaryAutomatically`. Not a showstopper, though.

File: ls_extension/main.go [hunk 9]
```diff
@@ -0,0 +1,105 @@
+/*
+ * © 2022 Snyk Limited All rights reserved.
```
Comment (bastiandoetsch) [inline ls_extension/main.go:2]: same :)

**Review by bastiandoetsch — APPROVED** [review]

**Pull request merged**